### PR TITLE
Parameter-Free Optimization Algorithms

### PR DESCRIPTION
**Pull request by Red-Portal**
This PR adds some recent parameter-free optimization algorithms. These algorithms should provide close-to-optimal performance without any tuning. The hope is to choose one of these algorithms as the default strategy so that `Turing` users don't need to perform any tuning when using `AdvancedVI`. In particular, the COCOB algorithm have been [reported](https://arxiv.org/abs/2301.11294) to be very effective for particle variational inference algorithms. 

The PR also adds parameter-averaging strategies. Some of the parameter-free algorithms report the best performance when combined with parameter averaging. [This paper](https://proceedings.neurips.cc/paper/2020/hash/7cac11e2f46ed46c339ec3d569853759-Abstract.html) also previously suggested that VI is best combined with parameter averaging (albeit by determining when to start averaging through the $\widehat{R}$ measure.)

Once the PosteriorDB project is done, we can probably run some large-scale experiments to determine which one is best.

- [x] Add parameter-free optimization algorithms
    - [x] DoG
    - [x] DoWG
    - [x] COCOB
 - [x] Parameter averaging strategies
     - [x] No averaging
     - [x] Polynomial averaging
     - [ ] (maybe?)  $\widehat{R}$-based adaptive averaging strategy
  - [x] Documentation
  - [x] Unit tests

### CODE REVIEW

**Comment by Red-Portal** [comment]
This PR is now ready. @yebai @mhauru @sunxd3 If anybody could take a look it would be great! Running the formatter unfortunately messed up the diff a little bit. Apologies for this in advance.

File: src/optimization/rules.jl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    repsilon = 1e-6
+end
+
+Optimisers.init(o::DoWG, x::AbstractArray{T}) where {T} = (copy(x), zero(T), T(o.repsilon))
+
+function Optimisers.apply!(::DoWG, state, x::AbstractArray{T}, dx) where {T}
+    x0, v, r = state
+
+    r = max(sqrt(sum(abs2, x - x0)), r)
+    r2 = r * r
```
[the remark this answers is not in the document]
Comment (Red-Portal) [reply src/optimization/rules.jl:24]: Wasn't there a recent trend in avoiding subscript/superscripts due to rendering issues on some platforms? @torfjelde any insight?

File: src/optimization/rules.jl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+Optimisers.@def struct DoWG <: Optimisers.AbstractRule
+    repsilon = 1e-6
+end
+
+Optimisers.init(o::DoWG, x::AbstractArray{T}) where {T} = (copy(x), zero(T), T(o.repsilon))
+
+function Optimisers.apply!(::DoWG, state, x::AbstractArray{T}, dx) where {T}
+    x0, v, r = state
+
+    r = max(sqrt(sum(abs2, x - x0)), r)
```
[the remark this answers is not in the document]
Comment (Red-Portal) [reply src/optimization/rules.jl:23]: I think `sum(abs2)` will be better, considering future GPU support.

File: src/optimize.jl [hunk 3]
```diff
@@ -16,26 +16,29 @@ This requires the variational approximation to be marked as a functor through `F
...
 The callback function `callback` has a signature of
 
-    callback(; stat, state, params, restructure, gradient)
+    callback(; stat, state, params, params_average, restructure, gradient)
 
 The arguments are as follows:
 - `stat`: Statistics gathered during the current iteration. The content will vary depending on `objective`.
 - `state`: Collection of the internal states used for optimization.
 - `params`: Variational parameters.
+- `params_average`: Variational parameters computed by the averaging strategy.
```
[the remark this answers is not in the document]
Comment (Red-Portal) [reply src/optimize.jl:41]: I agree. Let me update the code

File: src/optimization/rules.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhauru) [reply src/optimization/rules.jl:24]: My two cents is that using non-ASCII characters in code, especially library code, makes it harder to use command line tools like grep on the code files. Some editors also don't handle them fantastically.

**Pull request merged**